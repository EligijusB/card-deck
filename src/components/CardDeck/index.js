import { useCallback, useState } from "react";
import cardDeck from "../../mocks/card-deck";
import Card from "../Card";

import "./index.scss";

const CardDeck = () => {
    const [deck, setDeck] = useState(cardDeck);
    const [hand, setHand] = useState([]);

    const handleShuffle = useCallback(() => {
        const shuffledDeck = deck.slice();
        shuffledDeck.sort(() => Math.random() - 0.5);
        setDeck(shuffledDeck);
    }, [deck]);

    const handleDraw = useCallback(() => {
        let randomIndex = null;
        let card = null
        do {
            randomIndex = Math.floor(Math.random() * 51);
            card = deck[randomIndex];
        } while (card == null);

        setHand(hand.concat(card));
        const newDeck = deck.slice();
        newDeck.splice(randomIndex, 1);
        setDeck(newDeck);
    }, [hand, deck]);

    const handleReset = useCallback(() => {
        setHand([]);
        setDeck(cardDeck);
    }, []);

    const handleSort = useCallback(() => {
        const sortedDeck = deck.slice();
        sortedDeck.sort((a, b) => a.value === b.value ? a.suit.localeCompare(b.suit) : a.value - b.value);
        setDeck(sortedDeck);
    }, [deck]);

    return (
        <div className="card-deck" data-testid="card-deck">
            <div className="deck-actions">
                <div>
                    <button onClick={handleReset}>Reset</button>
                    <button onClick={handleSort}>Sort</button>
                    <button onClick={handleShuffle}>Shuffle</button>
                    <button onClick={handleDraw}>Draw</button>
                </div>
                {hand.length > 0 &&
                    <div data-testid="hand">
                        <div>Your hand:</div>
                        {hand.map(({ title, src }) => <Card key={title} title={title} src={src} />)}
                    </div>
                }
            </div>
            <div className="deck" data-testid="deck">
                {deck.map(({ title, src }) => <Card key={title} title={title} src={src} />)}
            </div>
        </div>
    );
};

export default CardDeck;