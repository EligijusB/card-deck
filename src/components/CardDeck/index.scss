.card-deck {
    display: grid;
    grid-template-columns: 1fr 3fr;

    @media (max-width: 768px) {
        grid-template-columns: 1fr 2fr;
    }

    @media (max-width: 600px) {
        grid-template-columns: 1fr 1fr;
    }

    .deck-actions {
        text-align: center;
        button {
            padding: 5px;
            margin: 5px;
        }
    }
}