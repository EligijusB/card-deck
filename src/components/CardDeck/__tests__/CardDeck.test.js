import { fireEvent, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import CardDeck from "..";

describe("<CardDeck />", () => {

    it("Should render correctly", () => {
        const comp = renderer.create(<CardDeck />);
        expect(comp.toJSON()).toMatchSnapshot();
    });

    it('Should allow user to sort and reset the deck', () => {
        render(<CardDeck />);
        const originalFirstCard = screen.getAllByRole('img')[0];

        fireEvent.click(screen.getByText("Sort"));
        let newFirstCard = screen.getAllByRole('img')[0];
        expect(originalFirstCard).not.toEqual(newFirstCard);

        fireEvent.click(screen.getByText("Reset"));
        newFirstCard = screen.getAllByRole('img')[0];
        expect(originalFirstCard).toEqual(newFirstCard);
    });

    it('Should allow user to shuffle', async () => {
        const {container} = render(<CardDeck />);
        const originalDeck = container.querySelector('[data-testid="deck"]');

        fireEvent.click(screen.getByText("Shuffle"));

        const newDeck = screen.getAllByRole('img')[0];
        expect(originalDeck).not.toEqual(newDeck);
    });

    it('Should allow user to draw', () => {
        render(<CardDeck />);

        expect(screen.queryByTestId("hand")).not.toBeInTheDocument();

        fireEvent.click(screen.getByText("Draw"));

        expect(screen.getByTestId("hand")).toBeInTheDocument();
    });

});