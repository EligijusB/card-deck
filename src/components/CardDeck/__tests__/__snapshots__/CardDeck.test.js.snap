// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardDeck /> Should render correctly 1`] = `
<div
  className="card-deck"
  data-testid="card-deck"
>
  <div
    className="deck-actions"
  >
    <div>
      <button
        onClick={[Function]}
      >
        Reset
      </button>
      <button
        onClick={[Function]}
      >
        Sort
      </button>
      <button
        onClick={[Function]}
      >
        Shuffle
      </button>
      <button
        onClick={[Function]}
      >
        Draw
      </button>
    </div>
  </div>
  <div
    className="deck"
    data-testid="deck"
  >
    <img
      alt="Ace of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/3/36/Playing_card_club_A.svg/1024px-Playing_card_club_A.svg.png"
    />
    <img
      alt="2 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/f/f5/Playing_card_club_2.svg/1024px-Playing_card_club_2.svg.png"
    />
    <img
      alt="3 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/6/6b/Playing_card_club_3.svg/1024px-Playing_card_club_3.svg.png"
    />
    <img
      alt="4 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/3/3d/Playing_card_club_4.svg/1024px-Playing_card_club_4.svg.png"
    />
    <img
      alt="5 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/5/50/Playing_card_club_5.svg/1024px-Playing_card_club_5.svg.png"
    />
    <img
      alt="6 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/a/a0/Playing_card_club_6.svg/1024px-Playing_card_club_6.svg.png"
    />
    <img
      alt="7 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/4/4b/Playing_card_club_7.svg/1024px-Playing_card_club_7.svg.png"
    />
    <img
      alt="8 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/e/eb/Playing_card_club_8.svg/1024px-Playing_card_club_8.svg.png"
    />
    <img
      alt="9 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/2/27/Playing_card_club_9.svg/1024px-Playing_card_club_9.svg.png"
    />
    <img
      alt="10 of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/3/3e/Playing_card_club_10.svg/1024px-Playing_card_club_10.svg.png"
    />
    <img
      alt="Jack of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/b/b7/Playing_card_club_J.svg/1024px-Playing_card_club_J.svg.png"
    />
    <img
      alt="Queen of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/f/f2/Playing_card_club_Q.svg/1024px-Playing_card_club_Q.svg.png"
    />
    <img
      alt="King of clubs"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/2/22/Playing_card_club_K.svg/1024px-Playing_card_club_K.svg.png"
    />
    <img
      alt="Ace of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/d/d3/Playing_card_diamond_A.svg/1024px-Playing_card_diamond_A.svg.png"
    />
    <img
      alt="2 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/5/59/Playing_card_diamond_2.svg/1024px-Playing_card_diamond_2.svg.png"
    />
    <img
      alt="3 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/8/82/Playing_card_diamond_3.svg/1024px-Playing_card_diamond_3.svg.png"
    />
    <img
      alt="4 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/2/20/Playing_card_diamond_4.svg/1024px-Playing_card_diamond_4.svg.png"
    />
    <img
      alt="5 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/f/fd/Playing_card_diamond_5.svg/1024px-Playing_card_diamond_5.svg.png"
    />
    <img
      alt="6 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/8/80/Playing_card_diamond_6.svg/1024px-Playing_card_diamond_6.svg.png"
    />
    <img
      alt="7 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/e/e6/Playing_card_diamond_7.svg/1024px-Playing_card_diamond_7.svg.png"
    />
    <img
      alt="8 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/7/78/Playing_card_diamond_8.svg/1024px-Playing_card_diamond_8.svg.png"
    />
    <img
      alt="9 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/9/9e/Playing_card_diamond_9.svg/1024px-Playing_card_diamond_9.svg.png"
    />
    <img
      alt="10 of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/3/34/Playing_card_diamond_10.svg/1024px-Playing_card_diamond_10.svg.png"
    />
    <img
      alt="Jack of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/a/af/Playing_card_diamond_J.svg/1024px-Playing_card_diamond_J.svg.png"
    />
    <img
      alt="Queen of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/0/0b/Playing_card_diamond_Q.svg/1024px-Playing_card_diamond_Q.svg.png"
    />
    <img
      alt="King of diamonds"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/7/78/Playing_card_diamond_K.svg/1024px-Playing_card_diamond_K.svg.png"
    />
    <img
      alt="Ace of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/5/57/Playing_card_heart_A.svg/1024px-Playing_card_heart_A.svg.png"
    />
    <img
      alt="2 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/d/d5/Playing_card_heart_2.svg/1024px-Playing_card_heart_2.svg.png"
    />
    <img
      alt="3 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Playing_card_heart_3.svg/1024px-Playing_card_heart_3.svg.png"
    />
    <img
      alt="4 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/a/a2/Playing_card_heart_4.svg/1024px-Playing_card_heart_4.svg.png"
    />
    <img
      alt="5 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/5/52/Playing_card_heart_5.svg/1024px-Playing_card_heart_5.svg.png"
    />
    <img
      alt="6 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/c/cd/Playing_card_heart_6.svg/1024px-Playing_card_heart_6.svg.png"
    />
    <img
      alt="7 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/9/94/Playing_card_heart_7.svg/1024px-Playing_card_heart_7.svg.png"
    />
    <img
      alt="8 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/5/50/Playing_card_heart_8.svg/1024px-Playing_card_heart_8.svg.png"
    />
    <img
      alt="9 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/5/50/Playing_card_heart_9.svg/1024px-Playing_card_heart_9.svg.png"
    />
    <img
      alt="10 of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/9/98/Playing_card_heart_10.svg/1024px-Playing_card_heart_10.svg.png"
    />
    <img
      alt="Jack of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/4/46/Playing_card_heart_J.svg/1024px-Playing_card_heart_J.svg.png"
    />
    <img
      alt="Queen of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/7/72/Playing_card_heart_Q.svg/1024px-Playing_card_heart_Q.svg.png"
    />
    <img
      alt="King of hearts"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/d/dc/Playing_card_heart_K.svg/1024px-Playing_card_heart_K.svg.png"
    />
    <img
      alt="Ace of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/2/25/Playing_card_spade_A.svg/1024px-Playing_card_spade_A.svg.png"
    />
    <img
      alt="2 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/f/f2/Playing_card_spade_2.svg/1024px-Playing_card_spade_2.svg.png"
    />
    <img
      alt="3 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/5/52/Playing_card_spade_3.svg/1024px-Playing_card_spade_3.svg.png"
    />
    <img
      alt="4 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/2/2c/Playing_card_spade_4.svg/1024px-Playing_card_spade_4.svg.png"
    />
    <img
      alt="5 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/9/94/Playing_card_spade_5.svg/1024px-Playing_card_spade_5.svg.png"
    />
    <img
      alt="6 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/d/d2/Playing_card_spade_6.svg/1024px-Playing_card_spade_6.svg.png"
    />
    <img
      alt="7 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/6/66/Playing_card_spade_7.svg/1024px-Playing_card_spade_7.svg.png"
    />
    <img
      alt="8 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/2/21/Playing_card_spade_8.svg/1024px-Playing_card_spade_8.svg.png"
    />
    <img
      alt="9 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/e/e0/Playing_card_spade_9.svg/1024px-Playing_card_spade_9.svg.png"
    />
    <img
      alt="10 of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/8/87/Playing_card_spade_10.svg/1024px-Playing_card_spade_10.svg.png"
    />
    <img
      alt="Jack of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/b/bd/Playing_card_spade_J.svg/1024px-Playing_card_spade_J.svg.png"
    />
    <img
      alt="Queen of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/5/51/Playing_card_spade_Q.svg/1024px-Playing_card_spade_Q.svg.png"
    />
    <img
      alt="King of spades"
      className="card"
      src="//upload.wikimedia.org/wikipedia/commons/thumb/9/9f/Playing_card_spade_K.svg/1024px-Playing_card_spade_K.svg.png"
    />
  </div>
</div>
`;
